packages/core/utils/src/product/csv-normalizer.ts

```diff
@@ -0,0 +1,518 @@
+      if (numericValue === undefined) {
+        throw createError(
+          rowNumber,
+          `Invalid value provided for "${columnName}". Expected value to be a number, received "${value}"`
+        )
+      } else {
+        output["prices"].push({
+          currency_code: iso,
+          amount: String(numericValue),
+        })
+      }
+    })
+  },
+}
+
+/**
+ * Options are processed separately and then defined on both the products and
+ * the variants.
+ */
+const optionColumns: {
+  [columnName: string]: ColumnProcessor<{
+    options: { key: any; value: any }[]
+  }>
+} = {
+  "Variant Option": (csvRow, rowColumns, rowNumber, output) => {
+    const matcher = /Variant Option \d+ Name/
+    const optionNameColumns = rowColumns.filter((rowKey) => {
+      return matcher.test(rowKey) && isPresent(normalizeValue(csvRow[rowKey]))
+    })
+
+    output["options"] = optionNameColumns.map((columnName) => {
+      const [, , counter] = columnName.split(" ")
+      const key = normalizeValue(csvRow[columnName])
+      const value = normalizeValue(csvRow[`Variant Option ${counter} Value`])
+
+      if (!isPresent(value)) {
+        throw createError(rowNumber, `Missing option value for "${columnName}"`)
+      }
+
+      return {
+        key,
+        value,
+      }
+    })
+  },
+}
+
+/**
+ * CSV normalizer processes all the allowed columns from a CSV file and remaps
+ * them into a new object with properties matching the "AdminCreateProduct".
+ *
+ * However, further validations must be performed to validate the format and
+ * the required fields in the normalized output.
+ */
+export class CSVNormalizer {
+  #rows: Record<string, string | boolean | number>[]
+
+  #products: {
+    toCreate: {
+      [handle: string]: {
+        [K in keyof AdminCreateProduct]?: any
+      }
+    }
+    toUpdate: {
+      [id: string]: {
+        [K in keyof AdminCreateProduct]?: any
+      }
+    }
+  } = {
+    toCreate: {},
+    toUpdate: {},
+  }
+
+  constructor(rows: Record<string, string | boolean | number>[]) {
+    this.#rows = rows
+  }
+
+  /**
+   * Ensures atleast one of the product id or the handle is provided. Otherwise
+   * we cannot process the row
+   */
+  #ensureRowHasProductIdentifier(
+    row: Record<string, string | boolean | number>,
+    rowNumber: number
+  ) {
+    const productId = row["Product Id"]
+    const productHandle = row["Product Handle"]
+    if (!isPresent(productId) && !isPresent(productHandle)) {
+      throw createError(
+        rowNumber,
+        "Missing product id and handle. One of them are required to process the row"
+      )
+    }
+
+    return { productId, productHandle }
+  }
+
+  /**
+   * Initializes a product object or returns an existing one
+   * by its id. The products with ids are treated as updates
+   */
+  #getOrInitializeProductById(id: string) {
+    if (!this.#products.toUpdate[id]) {
+      this.#products.toUpdate[id] = {
+        variants: [],
```

Comment:
Q: would it make sense to not initialize the variants here and instead below assign variants to `product['variants']`  directly? variants are optional if I am correct and in case you just want to update the product attribute I think providing an empty variant array would result in removing all existing variants but to verify

Reply:
Done